### PR TITLE
Target parallel strategy

### PR DESCRIPTION
**Pull request by yzh119**
This PR allows minigun to use target-parallel strategy. User could select to use Node Parallel (by source/target node) or Edge Parallel depend on how their functor works.

The PR is based on Lingfan's coo branch (use coo instead of csr in `CudaAdvanceAllGunrockLBOutKernel` to remove the overhead of binary search). It is still in progress and several examples need to be refactored.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,7 +17,7 @@ else()
 endif()
 
 minigun_option(USE_CUDA "Build with CUDA" ON)
-minigun_option(BUILD_SAMPLES "Build minigun samples" OFF)
+minigun_option(BUILD_SAMPLES "Build minigun samples" ON)
```
[the remark this answers is not in the document]
Comment (yzh119) [reply CMakeLists.txt]: Sorry I should not include this change in th PR.

File: minigun/minigun/cuda/advance_all.cuh [hunk 2]
```diff
@@ -36,17 +38,21 @@ template <typename Idx,
...
-    Idx dst = _ldg(csr.column_indices.data + eid);
+  while (eid < coo.column.length) {
+    Idx src = _ldg(coo.row.data + eid);
+    Idx dst = _ldg(coo.column.data + eid);
+    /* TODO(zihao): what is this?
+    if (src + dst < 0) {
+      // garbage code that cannot be optimized out but will never be executed
+      gdata.out[0] = -1;
+    }
+    */
```
[the remark this answers is not in the document]
Comment (yzh119) [reply minigun/minigun/cuda/advance_all.cuh]: I see, I just found there was a coo branch in your repo that has done something I did before, so I deprecated my own changes and turn to develop on this branch.

File: minigun/minigun/cuda/advance_all.cuh [hunk 3]
```diff
@@ -63,46 +43,143 @@ __global__ void CudaAdvanceAllGunrockLBOutKernel(
...
+              Functor::ApplyEdgeReduce(src, dst, eid, feat_idx, val, &gdata);
+          }
+          if (outbuf != nullptr)
+            outbuf[outoff] = val;
+          feat_idx += stride_x;
+        }
+      }
+      src += stride_y;
+    }
+  }
```
Comment (jermainewang) [inline minigun/minigun/cuda/advance_all.cuh]: Same. Try merge the two branches.

File: minigun/minigun/cuda/advance_all.cuh [hunk 4]
```diff
@@ -116,11 +193,24 @@ void CudaAdvanceAll(
         << " but only got a buffer of length " << output_frontier->length;
     }
   }
-  IntArray1D<Idx> outbuf = (output_frontier)? *output_frontier : IntArray1D<Idx>();
+   */
```
Comment (jermainewang) [inline minigun/minigun/cuda/advance_all.cuh]: remove the unused code

File: minigun/minigun/cuda/advance_all.cuh [hunk 5]
```diff
@@ -116,11 +193,24 @@ void CudaAdvanceAll(
...
+  IntArray1D<Idx> outbuf = IntArray1D<Idx>(); //(output_frontier)? *output_frontier : IntArray1D<Idx>();
```
Comment (jermainewang) [inline minigun/minigun/cuda/advance_all.cuh]: change back?

File: minigun/minigun/spmat.h [hunk 6]
```diff
@@ -27,6 +27,25 @@ struct Csr {
 using IntCsr = Csr<int32_t>;
 using LongCsr = Csr<int64_t>;
 
+template <typename Idx>
+struct Coo {
+  IntArray1D<Idx> row;
+  IntArray1D<Idx> column;
+};
```
Comment (jermainewang) [inline minigun/minigun/spmat.h]: Add `num_rows` and `num_cols` for both Csr and Coo?

**Comment by jermainewang** [comment]
Also, it will be great if we could write a programming guide. Probably after we cleanup the codes (e.g., remove the unused code for partial update) and implement the new block/warp-wise routines.

File: minigun/minigun/cpu/advance_all.h [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+void CPUAdvanceAllEdgeParallel(
+    const Coo<Idx>& coo,
+    GData *gdata) {
+  const Idx E = coo.column.length;
+#pragma omp parallel for
+  for (Idx eid = 0; eid < E; ++eid) {
+    const Idx src = coo.row.data[eid];
+    const Idx dst = coo.column.data[eid];
+    if (Functor::CondEdge(src, dst, eid, gdata))
+      Functor::ApplyEdge(src, dst, eid, gdata);
```
[the remark this answers is not in the document]
Comment (yzh119) [reply minigun/minigun/cpu/advance_all.h]: No, in edge parallel mode there is nothing to reduce/aggregate.

File: minigun/minigun/cpu/advance_all.h [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+              if (Functor::CondEdge(src, dst, eid, gdata)) {
+                Functor::ApplyEdgeReduce(src, dst, eid, feat_idx, val, gdata);
+              }
+            }
+            if (outbuf != nullptr)
+              outbuf[outoff] = val;
+          }
+        }
+      }
+    } else {
```
[the remark this answers is not in the document]
Comment (yzh119) [reply minigun/minigun/cpu/advance_all.h]: Let me have a try.

File: minigun/minigun/cpu/advance_all.h [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    const Csr<Idx>& csr,
+    GData *gdata) {
+  const Idx N = csr.row_offsets.length - 1;
+  const Idx feat_size = Functor::GetFeatSize(gdata);
+  DType *outbuf = Functor::GetOutBuf(gdata);
+#pragma omp parallel
+  {
+    DType val = static_cast<DType>(0);
+    if (Config::kParallel == kDst) {
+#pragma omp for
```
[the remark this answers is not in the document]
Comment (yzh119) [reply minigun/minigun/cpu/advance_all.h]: Yes, the problem is the definition of `val` should be inside `#pragma omp parallel`.
This caused a race condition, and it took me and @classicsong some time to debug.

File: minigun/minigun/cpu/advance_all.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (yzh119) [reply minigun/minigun/cpu/advance_all.h]: DONE

File: minigun/minigun/cuda/advance_all.cuh (hunk 4 above)
Comment (yzh119) [reply minigun/minigun/cuda/advance_all.cuh]: DONE

File: minigun/minigun/cuda/advance_all.cuh (hunk 3 above)
Comment (yzh119) [reply minigun/minigun/cuda/advance_all.cuh]: DONE

File: minigun/minigun/cpu/advance_all.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (yzh119) [reply minigun/minigun/cpu/advance_all.h]: DONE

File: minigun/minigun/cuda/advance_all.cuh (hunk 5 above)
Comment (yzh119) [reply minigun/minigun/cuda/advance_all.cuh]: removed `output_frontier` in AdvanceAll.

File: minigun/minigun/spmat.h (hunk 6 above)
Comment (yzh119) [reply minigun/minigun/spmat.h]: DONE.

**Comment by yzh119** [comment]
Yes I'll draft a programming guide in the next PR.

**Pull request merged**